test/over_react/util/safe_render_manager/safe_render_manager_test.dart

```diff
@@ -0,0 +1,707 @@
+      });
+
+      test('rerenders a component into the specified `mountNode`', () {
+        renderManager.render(Dom.div()('foo'));
+        renderManager.render(Dom.div()('bar'));
+
+        expect(mountNode.text, 'bar');
+      });
+
+      group('renders a component and exposes a ref to it via `contentRef`', () {
+        test('when there is no existing ref', () {
+          renderManager.render(Wrapper()());
+
+          expect(renderManager.contentRef, isNotNull);
+          expect(renderManager.contentRef, const isInstanceOf<WrapperComponent>());
+        });
+
+        test('by chaining any existing callback ref', () {
+          WrapperComponent existingWrapperRef;
+
+          renderManager.render((Wrapper()..ref = ((ref) => existingWrapperRef = ref))());
+
+          expect(renderManager.contentRef, isNotNull);
+          expect(existingWrapperRef, same(renderManager.contentRef));
+        });
+      });
+    });
+
+    group('tryUnmount()', () {
+      group('unmounts the rendered component', () {
+        setUp(() {
+          renderManager.render(Wrapper()());
+          expect(mountNode.children, isNotEmpty);
+        });
+
+        test('', () async {
+          renderManager.tryUnmount();
+          expect(mountNode.children, isEmpty);
+        });
+
+        test('and calls the provided callback when complete', () async {
+          renderManager.render(Wrapper()());
+
+          expect(mountNode.children, isNotEmpty);
+
+          final onUnmountCompleter = new Completer();
+
+          renderManager.tryUnmount(onMaybeUnmounted: bind1Guarded((isUnmounted) {
+            expect(mountNode.children, isEmpty);
+            expect(isUnmounted, isTrue);
+
+            onUnmountCompleter.complete();
+          }));
+
+          await onUnmountCompleter.future;
+        });
+      });
+
+      test('invokes the provided callback immediately when nothing has been rendered', () async {
+        expect(mountNode.children, isEmpty, reason: 'test setup sanity check');
+
+        bool onUnmountCalledSynchronously = false;
+
+        renderManager.tryUnmount(onMaybeUnmounted: bind1Guarded((isUnmounted) {
+          expect(isUnmounted, isTrue);
+          onUnmountCalledSynchronously = true;
+        }));
+
+        expect(onUnmountCalledSynchronously, isTrue);
+      });
+    });
+
+    group('automatically attaches and detached the mount node', () {
+      setUp(() async {
+        // Clean up the manager from the above setUp block.
+        await renderManager?.dispose();
+        renderManager = null;
+      });
+
+      test('when autoAttachMountNode is true', () {
+        renderManager = new SafeRenderManager(autoAttachMountNode: true);
+        expect(document.contains(renderManager.mountNode), isFalse, reason: 'test setup check');
+
+        renderManager.render(Dom.div()());
+        expect(renderManager.mountNode.parent, document.body);
+
+        renderManager.tryUnmount();
+        expect(renderManager.mountNode.parent, isNull);
```

Comment:
I realize this is probably kosher - but when I first read this - I thought... _it... null'd `document.body`??!!?!!?_

Not saying you have to change it, but something like this would test the same thing:

```dart
expect(document.body.contains(renderManager.mountNode), isFalse);
```

Reply:
I actually had this exact thing at first, but then changed it to what I have now so that the failure messages would be more helpful.

> Expected: null
> Actual: BodyElement

vs

> Expected: false 
> Actual: true